pyteal/ast/jsonref.py

```diff
@@ -0,0 +1,77 @@
+from typing import TYPE_CHECKING
+from enum import Enum
+
+from pyteal.types import TealType, require_type
+from pyteal.errors import TealTypeError, verifyFieldVersion, verifyTealVersion
+from pyteal.ir import TealOp, Op, TealBlock
+from pyteal.ast.expr import Expr
+from pyteal.ast.leafexpr import LeafExpr
+
+if TYPE_CHECKING:
+    from pyteal.compiler import CompileOptions
+
+
+class JsonRefType(Enum):
+    # fmt: off
+    #              id |    name   |      type      | min version
+    json_string = (0, "JSONString", TealType.bytes,  7)
+    json_uint64 = (1, "JSONUint64", TealType.uint64, 7)
+    json_object = (2, "JSONObject", TealType.bytes,  7)
+    # fmt: on
+
+    def __init__(self, id: int, name: str, type: TealType, min_version: int) -> None:
+        self.id = id
+        self.arg_name = name
+        self.ret_type = type
+        self.min_version = min_version
+
+    def type_of(self) -> TealType:
+        return self.ret_type
+
+
+JsonRefType.__module__ = "pyteal"
+
+
+class JsonRef(LeafExpr):
+    """An expression that accesses a key from a supported utf-8 encoded json object.
+
+    The json object must satisfy a [particular specification](https://github.com/algorand/go-algorand/blob/master/data/transactions/logic/jsonspec.md).
+    """
+
+    json_string = JsonRefType.json_string
+    json_uint64 = JsonRefType.json_uint64
+    json_object = JsonRefType.json_object
+
+    def __init__(self, type: JsonRefType, json_obj: Expr, key: Expr) -> None:
+        super().__init__()
+
+        if not isinstance(type, JsonRefType):
+            raise TealTypeError(type, JsonRefType)
+        self.type = type
+
+        require_type(json_obj, TealType.bytes)
+        self.json_obj = json_obj
+
+        require_type(key, TealType.bytes)
+        self.key = key
+
+    def __teal__(self, options: "CompileOptions"):
+        verifyTealVersion(
+            Op.json_ref.min_version,
+            options.version,
+            "TEAL version too low to use op json_ref",
+        )
+
+        verifyFieldVersion(self.type.arg_name, self.type.min_version, options.version)
+
+        op = TealOp(self, Op.json_ref, self.type.arg_name)
+        return TealBlock.FromOp(options, op, self.json_obj, self.key)
+
+    def __str__(self):
+        return "(JsonRef {})".format(self.type.arg_name)
+
+    def type_of(self):
+        return self.type.type_of()
+
```

Comment:
I like how you had the base64 `.std/.url`, does it make sense to do something similar here?

Reply:
It very well might. I could easily be convinced. My thinking is that if I saw the following:
```py
JsonRef.string(Bytes('{"a":"b"}'), Bytes("a"))
```
I wouldn't be sure what the `.string` property meant at first glance. Without a thorough understanding of the `json_ref` opcode I wouldn't immediately understand whether string refers to the key type, or the input type, or the value's decoded type.

My hope was that explicitly having to set the format would make it clearer at first glance.

```py
JsonRef(JsonRef.json_string, Bytes('{"a":"b"}'), Bytes("a"))
```

However, as I'm writing this, I'm thinking that named functions `.as_string`, `.as_uint64`, and `.as_object` would solve both issues. I'll add those 👍

Reply:
@barnjamin I'm having second thoughts about this. As I work more with the PyTeal library, it seems that the "PyTeal way" of doing things is to use dot notation for accessing _properties_ but passed parameters for encodings or specifying return value types. For example, you see `AssetParam.total()` to access a property of an asset param, but ` EcdsaVerify(EcdsaCurve.Secp256k1, *args)` for ecdsa verification.

I actually think using dot notation for everything could provide a better developer experience and would probably be in favor of adopting that method moving forward. However, if we were trying to stay consistent with existing behavior, we should get rid of `.std/.url/.as_string/etc` in favor of exported enums as function parameters.

Reply:
My lightly held 2c:  I'm on-board with the PR's approach.  If another reviewer has a strong opinion otherwise, I can be convinced to revert for consistency.